source/evp_hash.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright 2025 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * This CLI tool computes hashes using the specified algorithm. Uses the EVP API
+ * by default, but this tool can also use the corresponding deprecated API's.
+ * Prints out the average time per hash computation.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#ifndef _WIN32
+# include <unistd.h>
+# include <libgen.h>
+#else
+# include <windows.h>
+# include "perflib/getopt.h"
+#endif	/* _WIN32 */
+
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include "perflib/perflib.h"
+
+#define RUN_TIME 5
+#define DATA_SIZE 1500
+
+size_t *counts = NULL;
+OSSL_TIME max_time;
+int err = 0;
+static int threadcount;
+size_t num_calls;
+
+typedef enum {
+    SHA1_ALG = 0,
+    SHA224_ALG,
+    SHA256_ALG,
+    SHA384_ALG,
+    SHA512_ALG,
+} hash_algorithm_type;
+
+static unsigned char data[DATA_SIZE];
+static int deprecated_api = 0;
+static int update_times = 1;
+static int hash_algorithm = -1;
+
+int hash_deprecated()
+{
+    int i;
+    SHA_CTX sha_ctx;
+    SHA256_CTX sha256_ctx;
+    SHA512_CTX sha512_ctx;
+    unsigned char md[EVP_MAX_MD_SIZE];
+
+    switch (hash_algorithm) {
```

Comment:
I'd like to see this decision to go outside the critical section as it will add (even if little) overhead for the measurement.